text/0517-io-os-reform.md

```diff
@@ -483,7 +483,72 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::env`
 [std::env]: #stdenv
 
-> To be added in a follow-up PR.
+Most of what's available in `std::os` today will move to `std::env`,
+and the signatures will be updated to follow this RFC's
+[Design principles] as follows.
+
+**Arguments**:
+
+* `args`: change to yield an iterator rather than vector if possible; in any case, it should produce an `OsStrBuf`.
```

Comment:
I've investigated this a bit more, and here's what I've come up with:

* On Windows, arguments need to be parsed by [`CommandLineToArgvW`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb776391%28v=vs.85%29.aspx) which creates a local allocation which then needs to be freed by [`LocalFree`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa366730%28v=vs.85%29.aspx). This causes the data to unfortunately not have a static lifetime.
* On linux (and freebsd I believe) we currently don't have implemented an easy way to fetch argc/argv from the system, so we squirrel them away ourselves. Our own storage does not necessarily have a static lifetime because we must free it at some point
* On other platforms we use various per-platform support to load the args, but this is a global resource which can be modified by anyone.

As a result, I don't think that we can return an iterator over slices into static memory across all platforms, so we'll need to return an owned allocation of one kind or another.

In terms of whether this is an iterator or not, it turns out that on Windows it's actually quite natural to return an iterator (you have an owned chunk of memory you iterate over and then drop at the end). On unix it's not quite so natural, but certainly possible! (using `vec::IntoIter`).

In general we probably don't need to worry about optimizing this *too* much as I doubt `os::args` is in anyone's tight inner loop. I think that this is largely an API question and by returning an iterator structure we in theory open up ourselves to possible other optimizations (or options) in the future. The downside is that `os::args()[1]` (accessing a particular argument) is a little less ergonomic as it must be `.collect()`'d first.